api/utils/keys/yubikey.go

```diff
@@ -648,146 +738,137 @@ func (y *YubiKey) open() (yk *piv.YubiKey, err error) {
 	} else if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return yk, nil
-}
 
-func isRetryError(err error) bool {
-	const retryError = "connecting to smart card: the smart card cannot be accessed because of other connections outstanding"
-	return strings.Contains(err.Error(), retryError)
+	c.waitClose.Add(1)
+	go func() {
+		c.waitClose.Wait()
+
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.conn.Close()
+		c.conn = nil
```

Comment:
Hmm I think there may be a race condition here

1. all current connections expire and c.waitClose.Wait() returns
2. this func tries to lock the mutex
3. there is a concurrent call to connect() which locks the mutex first, adds 1 to the waitgroup, unlocks the mutex and returns
4. then this func gets the lock, closes the conn and sets it to nil
5. the other caller of connect probably panics trying to use the nil connection

Could probably solve it a couple ways, maybe maintain a count of open connections protected by the mutex and loop here while it's not 0? something like
```go
go func() {
	for {
		c.waitClose.Wait()
		c.mu.Lock()
		if c.openConnections == 0 {
			defer c.mu.Unlock()
			break
		}
		c.mu.Unlock()
	}
	c.conn.Close()
	c.conn = nil
}()
```

Reply:
Yep I found the same thing - https://github.com/gravitational/teleport/pull/47091#discussion_r1796015833

Reply:
Thanks, I used the Brian's code, should be fixed now.